File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        l, r = 0, 0
        lon = 0
        seen = set()
        while r < n:
            if s[r] in seen:
                while s[r] in seen:
                    seen.discard(s[l])
                    l += 1
            seen.add(s[r])
            lon = max(lon, r-l+1)
            r += 1
        return lon
            
                
        

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        int res = 0, n = s.length(), l = 0;

        Map<Character, Integer> m = new HashMap<>();


        for (int r = 0; r < n; r++) {
            Character c = s.charAt(r);

            if (m.containsKey(c)) {
                l = Math.max(l, m.get(c) + 1);
            }

            m.put(c, r);


            res = Math.max(res, (r-l + 1));
        }

        return res;
        
    }
}